File: src/test/java/space/npstr/wolfia/webapi/TogglzEndpointTest.java
/*
 * Copyright (C) 2016-2020 the original author or authors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package space.npstr.wolfia.webapi;

import java.io.IOException;
import java.util.Arrays;
import java.util.Base64;
import java.util.Set;
import java.util.stream.Collectors;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.context.SecurityContext;
import org.springframework.security.core.context.SecurityContextImpl;
import org.springframework.security.core.userdetails.User;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.session.Session;
import org.springframework.session.SessionRepository;
import space.npstr.wolfia.ApplicationTest;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * The Togglz Console is served by a Servlet, so we can't use MockMvc.
 */
class TogglzEndpointTest<T extends Session> extends ApplicationTest {

    private static final String TOGGLZ_PATH = "/api/togglz/index";

    private OkHttpClient httpClient;

    @Autowired
    private OkHttpClient.Builder httpClientBuilder;

    @Autowired
    private SessionRepository<T> sessionRepository;

    @BeforeEach
    void setup() {
        httpClient = httpClientBuilder
                .followRedirects(false)
                .build();
    }

    @Test
    void whenGet_withoutAuthentication_returnUnauthorized() throws Exception {
        Request request = getTogglzConsole()
                .build();
        Response response = httpClient.newCall(request).execute();

        assertThat(response.code()).isEqualTo(HttpStatus.FORBIDDEN.value());
    }

    @Test
    void whenGet_withUserAuthority_returnUnauthorized() throws Exception {
        Session session = generateHttpSession(Authorization.ROLE_USER);
        Request request = getTogglzConsole()
                .header(HttpHeaders.COOKIE, getSessionCookie(session))
                .build();

        Response response = httpClient.newCall(request).execute();

        assertThat(response.code()).isEqualTo(HttpStatus.FORBIDDEN.value());
    }

    @Test
    void whenGet_withOwnerAuthority_returnOk() throws IOException {
        Session session = generateHttpSession(Authorization.ROLE_OWNER);
        Request request = getTogglzConsole()
                .header(HttpHeaders.COOKIE, getSessionCookie(session))
                .build();

        Response response = httpClient.newCall(request).execute();

        assertThat(response.code()).isEqualTo(HttpStatus.OK.value());
    }

    private Request.Builder getTogglzConsole() {
        return new Request.Builder()
                .get()
                .url("http://localhost:" + port + TOGGLZ_PATH);
    }

    private String getSessionCookie(Session session) {
        return "SESSION=" + Base64.getEncoder().encodeToString(session.getId().getBytes());
    }

    private T generateHttpSession(final String... requestedAuthorities) {
        final Set<GrantedAuthority> authorities =
                Arrays.stream(requestedAuthorities)
                        .map(SimpleGrantedAuthority::new)
                        .collect(Collectors.toSet());

        final UserDetails userDetails = new User(
                "foo",
                "bar",
                true,
                true,
                true,
                true,
                authorities
        );

        final Authentication authentication = new UsernamePasswordAuthenticationToken(
                userDetails, userDetails.getPassword(), userDetails.getAuthorities());

        final UsernamePasswordAuthenticationToken authenticationToken = new UsernamePasswordAuthenticationToken(
                userDetails, authentication.getCredentials(), userDetails.getAuthorities());
        authenticationToken.setDetails(authentication.getDetails());

        final SecurityContext securityContext = new SecurityContextImpl(authentication);

        T session = sessionRepository.createSession();
        session.setAttribute("SPRING_SECURITY_CONTEXT", securityContext);
        session.setAttribute("sessionId", session.getId());
        sessionRepository.save(session);
        return session;
    }
}

File: src/test/java/space/npstr/wolfia/webapi/TogglzEndpointTest.kt
/*
 * Copyright (C) 2016-2025 the original author or authors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package space.npstr.wolfia.webapi

import java.util.Base64
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpMethod
import org.springframework.http.HttpStatus
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.context.SecurityContext
import org.springframework.security.core.context.SecurityContextImpl
import org.springframework.security.core.userdetails.User
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.session.Session
import org.springframework.session.SessionRepository
import space.npstr.wolfia.ApplicationTest

/**
 * The Togglz Console is served by a Servlet, so we can't use MockMvc.
 */
internal class TogglzEndpointTest<T : Session> : ApplicationTest() {

	private final val togglzConsolePath = "/api/togglz/index"

	@Autowired
	@Suppress("SpringJavaInjectionPointsAutowiringInspection")
	private lateinit var sessionRepository: SessionRepository<T>

	@Test
	fun whenGet_withoutAuthentication_returnUnauthorized() {
		val response = restTemplate.getForEntity("/$togglzConsolePath", Void::class.java)

		assertThat(response.statusCode).isEqualTo(HttpStatus.FORBIDDEN)
	}

	@Test
	fun whenGet_withUserAuthority_returnUnauthorized() {
		val headers = HttpHeaders()
		headers.add(HttpHeaders.COOKIE, sessionCookie(generateHttpSession(Authorization.USER)))

		val response = restTemplate.exchange(
			"/$togglzConsolePath",
			HttpMethod.GET,
			HttpEntity<Void>(headers),
			Void::class.java
		)

		assertThat(response.statusCode).isEqualTo(HttpStatus.FORBIDDEN)
	}

	@Test
	fun whenGet_withOwnerAuthority_returnOk() {
		val headers = HttpHeaders()
		headers.add(HttpHeaders.COOKIE, sessionCookie(generateHttpSession(Authorization.OWNER)))

		val response = restTemplate.exchange(
			"/$togglzConsolePath",
			HttpMethod.GET,
			HttpEntity<Void>(headers),
			Void::class.java
		)

		assertThat(response.statusCode).isEqualTo(HttpStatus.OK)
	}

	private fun sessionCookie(session: Session): String {
		return "SESSION=" + Base64.getEncoder().encodeToString(session.id.toByteArray())
	}

	private fun generateHttpSession(vararg requestedAuthorities: GrantedAuthority): T {
		val userDetails: UserDetails = User(
			"foo",
			"bar",
			true,
			true,
			true,
			true,
			requestedAuthorities.toSet()
		)
		val authentication: Authentication = UsernamePasswordAuthenticationToken(
			userDetails, userDetails.password, userDetails.authorities
		)
		val authenticationToken = UsernamePasswordAuthenticationToken(
			userDetails, authentication.credentials, userDetails.authorities
		)
		authenticationToken.details = authentication.details
		val securityContext: SecurityContext = SecurityContextImpl(authentication)
		val session = sessionRepository.createSession()!!
		session.setAttribute("SPRING_SECURITY_CONTEXT", securityContext)
		session.setAttribute("sessionId", session.id)
		sessionRepository.save(session)
		return session
	}
}
